digitalocean/datasource_digitalocean_image.go

```diff
@@ -53,42 +63,82 @@ func dataSourceDigitalOceanImage() *schema.Resource {
 func dataSourceDigitalOceanImageRead(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*godo.Client)
 
-	opts := &godo.ListOptions{}
+	opts := &godo.ListOptions{PerPage: 200}
+
+	name, nameOk := d.GetOk("name")
+	nameRegex, nameRegexOk := d.GetOk("name_regex")
 
-	images, _, err := client.Images.ListUser(context.Background(), opts)
-	if err != nil {
-		d.SetId("")
-		return err
+	if !nameOk && !nameRegexOk {
+		return fmt.Errorf("One of name or name_regex must be assigned")
 	}
-	image, err := findImageByName(images, d.Get("name").(string))
 
-	if err != nil {
-		return err
+	var match func(string) bool
+
+	if nameRegexOk {
+		match = regexp.MustCompile(nameRegex.(string)).MatchString
+	} else {
+		match = func(s string) bool {
+			return s == name.(string)
+		}
 	}
 
-	d.SetId(image.Name)
-	d.Set("name", image.Name)
-	d.Set("image", strconv.Itoa(image.ID))
-	d.Set("min_disk_size", image.MinDiskSize)
-	d.Set("private", !image.Public)
-	d.Set("regions", image.Regions)
-	d.Set("type", image.Type)
+	source := d.Get("source").(string)
 
-	return nil
-}
+	var listFn func(ctx context.Context, opt *godo.ListOptions) ([]godo.Image, *godo.Response, error)
+	switch source {
+	case "user":
+		listFn = client.Images.ListUser
+	case "application":
+		listFn = client.Images.ListApplication
+	case "distribution":
+		listFn = client.Images.ListDistribution
+	default:
+		return fmt.Errorf("source must be one of user, application, or distribution")
+	}
+
+	var images []*godo.Image
+
+	for {
+		images, res, err := listFn(context.Background(), opts)
```

Comment:
You are right! This was a mistake on my end, I'm shadowing the outer `images` var. I'll rename the outer variable.

Reply:
> I can confirm that the images are actually created, but it does not find it. It looks like this might be due to the scope of the `images` variable.

I think you are totally right.